crates/pico_macros/src/memo.rs

```diff
@@ -160,3 +194,76 @@ fn hash(input: &Signature) -> u64 {
     input.to_token_stream().to_string().hash(&mut s);
     s.finish()
 }
+
+enum ArgType {
+    Source,
+    MemoRef,
+    Other,
+}
+
+impl ArgType {
+    pub fn parse(ty: &syn::Type) -> Self {
+        if type_is(ty, "SourceId") {
+            return ArgType::Source;
+        }
+        if type_is(ty, "MemoRef") {
+            return ArgType::MemoRef;
+        }
+        ArgType::Other
+    }
+}
+
+fn type_is(ty: &syn::Type, target: &'static str) -> bool {
+    let inner = if let syn::Type::Reference(r) = ty {
+        &*r.elem
+    } else {
+        ty
+    };
+
+    if let syn::Type::Path(type_path) = inner {
+        if let Some(segment) = type_path.path.segments.last() {
+            return segment.ident == target;
+        }
+    }
+    false
+}
+
+fn get_fn_lifetime(sig: &mut Signature) -> Lifetime {
+    let lt = sig.generics.params.iter().find_map(|param| {
+        if let GenericParam::Lifetime(lt) = param {
+            Some(lt.lifetime.clone())
+        } else {
+            None
+        }
+    });
+    if lt.is_none() {
+        sig.generics.params.push(parse_quote!('db));
+    }
+    lt.unwrap_or(parse_quote!('db))
+}
+
+fn check_db_lifetime(arg: &mut FnArg, lifetime: Lifetime) -> Result<(), syn::Error> {
+    if let FnArg::Typed(PatType { ty, .. }) = arg {
+        if let syn::Type::Reference(ref mut reference) = **ty {
+            if let Some(db_lifetime) = &reference.lifetime {
+                if db_lifetime != &lifetime {
```

Comment:
Should this fail to compile?

```
fn parse_ast<'foo>(db: &Database, id: SourceId<Input>, bar: &'foo ()) -> Result<Program> {
```

it does not

Reply:
Macro assigns `'foo` to the Database in that case. Tbh I'm not sure what else we need to cover. My case is I need the same lifetime for function itself, db and MemoRef. I'll add `'db` or take the first lifetime from the function and assign it to the db (or check if this lifetime has been assigned by a user)